chasm/lib/workersession/statemachine.go

```diff
@@ -0,0 +1,90 @@
+package workersession
+
+import (
+	"time"
+
+	"go.temporal.io/server/chasm"
+	workersessionpb "go.temporal.io/server/chasm/lib/workersession/gen/workersessionpb/v1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	// Time to wait before cleaning up an expired session.
+	expiredSessionCleanupDelay = 10 * time.Minute
+)
+
+// RecordHeartbeat processes a heartbeat by applying the heartbeat received transition.
+func RecordHeartbeat(ctx chasm.MutableContext, ws *WorkerSession, leaseDeadline time.Time) error {
+	return TransitionHeartbeatReceived.Apply(ctx, ws, EventHeartbeatReceived{
+		Time:          time.Now(),
+		LeaseDeadline: leaseDeadline,
+	})
+}
+
+// scheduleLeaseExpiry schedules a timer task that will fire when the lease expires.
+func scheduleLeaseExpiry(ctx chasm.MutableContext, ws *WorkerSession, leaseDeadline time.Time) {
+	expiryTask := &workersessionpb.LeaseExpiryTask{
+		LeaseExpirationTime: timestamppb.New(leaseDeadline),
+	}
+
+	taskAttrs := chasm.TaskAttributes{
+		ScheduledTime: leaseDeadline,
+		Destination:   "", // Empty means local execution.
+	}
+
+	ctx.AddTask(ws, taskAttrs, expiryTask)
+}
+
+// EventHeartbeatReceived is triggered when a heartbeat is received from the worker.
+type EventHeartbeatReceived struct {
+	Time          time.Time
+	LeaseDeadline time.Time
+}
+
+// TransitionHeartbeatReceived handles heartbeat reception, extending the lease and scheduling a new timeout.
+var TransitionHeartbeatReceived = chasm.NewTransition(
+	[]workersessionpb.WorkerSessionStatus{workersessionpb.WORKER_SESSION_STATUS_ACTIVE},
+	workersessionpb.WORKER_SESSION_STATUS_ACTIVE,
```

Comment:
I wonder if we could transition back from expired to active but what you have now is fine IMHO.